src/Haskell/Ide/Engine/Plugin/Brittany.hs

```diff
@@ -20,52 +17,62 @@ import qualified Language.Haskell.LSP.Types.Lens       as J
 import           System.FilePath (FilePath, takeDirectory)
 import           Data.Maybe (maybeToList)
 
-data FormatParams = FormatParams Int Uri (Maybe Range)
-     deriving (Eq, Show, Generic, FromJSON, ToJSON)
-
 brittanyDescriptor :: PluginId -> PluginDescriptor
 brittanyDescriptor plId = PluginDescriptor
-  { pluginId       = plId
-  , pluginName     = "Brittany"
-  , pluginDesc     = "Brittany is a tool to format source code."
-  , pluginCommands = []
+  { pluginId                 = plId
+  , pluginName               = "Brittany"
+  , pluginDesc               = "Brittany is a tool to format source code."
+  , pluginCommands           = []
   , pluginCodeActionProvider = Nothing
   , pluginDiagnosticProvider = Nothing
-  , pluginHoverProvider = Nothing
-  , pluginSymbolProvider = Nothing
+  , pluginHoverProvider      = Nothing
+  , pluginSymbolProvider     = Nothing
   , pluginFormattingProvider = Just provider
   }
 
 -- | Formatter provider of Brittany.
 -- Formats the given source in either a given Range or the whole Document.
 -- If the provider fails an error is returned that can be displayed to the user.
-provider :: FormattingProvider
-provider uri formatType opts = pluginGetFile "brittanyCmd: " uri $ \file -> do
-  confFile <- liftIO $ getConfFile file
-  mtext <- readVFS uri
-  case mtext of
-    Nothing -> return $ IdeResultFail (IdeError InternalError "File was not open" Null)
-    Just text -> case formatType of
-      FormatRange r -> do
-        res <- liftIO $ runBrittany tabSize confFile $ extractRange r text
-        case res of
-          Left err -> return $ IdeResultFail (IdeError PluginError
-                      (T.pack $ "brittanyCmd: " ++ unlines (map showErr err)) Null)
-          Right newText -> do
-            let textEdit = J.TextEdit (normalize r) newText
-            return $ IdeResultOk [textEdit]
-      FormatDocument -> do
-        res <- liftIO $ runBrittany tabSize confFile text
-        case res of
-          Left err -> return $ IdeResultFail (IdeError PluginError
-                      (T.pack $ "brittanyCmd: " ++ unlines (map showErr err)) Null)
-          Right newText ->
-            return $ IdeResultOk [J.TextEdit (fullRange text) newText]
+provider
+  :: MonadIO m
+  => Text
+  -> Uri
+  -> FormattingType
+  -> FormattingOptions
+  -> m (IdeResult [TextEdit])
```

Comment:
Do you use this at another type of than `IdeM`?

Reply:
No. It is an abstraction, so it might be used in another type, but not yet. Afaik, by using another Monad, it might be possible to either wait for a response or to get it immediately. 
If this is bad style, we can drop it, though.